File: packages/swift/barnard/Sources/Barnard/BarnardEventInfo.swift
// Use of this source code is governed by a BSD-style license.

import Foundation

/// The unauthenticated, event-scoped hint carried by B005. It contains no
/// device, organizer, or other persistent identifier.
public struct BarnardEventInfo: Equatable {
  public let eventDisplayName: String
  public let eventCodeHash: Data
  /// Reserved for the optional `0x10` census extension. v1 leaves it absent.
  public let census: Data?

  public init(eventDisplayName: String, eventCodeHash: Data, census: Data? = nil) {
    self.eventDisplayName = eventDisplayName
    self.eventCodeHash = eventCodeHash
    self.census = census
  }
}

/// Local-only authorization state for serving B005. Neither flag is encoded
/// in the characteristic value and both deliberately default to `false`.
public struct BarnardEventInfoServePolicy: Equatable {
  public var organizerDesignated: Bool
  public var eventActiveForDiscovery: Bool

  public init(organizerDesignated: Bool = false, eventActiveForDiscovery: Bool = false) {
    self.organizerDesignated = organizerDesignated
    self.eventActiveForDiscovery = eventActiveForDiscovery
  }

  var mayServe: Bool { organizerDesignated && eventActiveForDiscovery }
}

public enum BarnardEventInfoError: Error, Equatable {
  case invalidPayloadLength
  case unsupportedFormatVersion
  case truncatedTlv
  case zeroTlvType
  case unorderedTlvTypes
  case missingDisplayName
  case missingEventCodeHash
  case invalidDisplayName
  case invalidEventCodeHash
  case eventCodeHashMismatch
}

/// B005's pure canonical encoder and strict parser. The encoder intentionally
/// delegates its hash derivation to the existing B004 implementation.
public enum BarnardEventInfoCodec {
  public static let formatVersion: UInt8 = 1
  public static let maximumPayloadBytes = 512
  public static let maximumDisplayNameBytes = 64

  public static func payloadIfServing(
    policy: BarnardEventInfoServePolicy,
    eventCode: String?,
    eventDisplayName: String?,
    b004EventCodeHash: Data
  ) throws -> Data? {
    guard policy.mayServe,
      let eventCode,
      !eventCode.isEmpty,
      let eventDisplayName
    else { return nil }
    return try serialize(
      eventCode: eventCode,
      eventDisplayName: eventDisplayName,
      b004EventCodeHash: b004EventCodeHash
    )
  }

  public static func serialize(
    eventCode: String,
    eventDisplayName: String,
    b004EventCodeHash: Data
  ) throws -> Data {
    let displayNameBytes = try canonicalDisplayNameBytes(eventDisplayName)
    let eventCodeHash = BarnardCrypto.computeEventCodeHash(eventCode)
    guard eventCodeHash.count == 8, eventCodeHash == b004EventCodeHash else {
      throw BarnardEventInfoError.eventCodeHashMismatch
    }

    var payload = Data([formatVersion, 0x01])
    appendLength(displayNameBytes.count, to: &payload)
    payload.append(displayNameBytes)
    payload.append(0x02)
    appendLength(eventCodeHash.count, to: &payload)
    payload.append(eventCodeHash)
    guard (16...maximumPayloadBytes).contains(payload.count) else {
      throw BarnardEventInfoError.invalidPayloadLength
    }
    return payload
  }

  public static func validateEventDisplayName(_ eventDisplayName: String) throws {
    _ = try canonicalDisplayNameBytes(eventDisplayName)
  }

  public static func parse(_ input: Data) throws -> BarnardEventInfo {
    let payload = Data(input)
    guard (16...maximumPayloadBytes).contains(payload.count) else {
      throw BarnardEventInfoError.invalidPayloadLength
    }
    guard payload.first == formatVersion else {
      throw BarnardEventInfoError.unsupportedFormatVersion
    }

    var index = 1
    var previousType = 0
    var displayName: String?
    var eventCodeHash: Data?
    while index < payload.count {
      guard index + 3 <= payload.count else { throw BarnardEventInfoError.truncatedTlv }
      let type = Int(payload[index])
      let length = (Int(payload[index + 1]) << 8) | Int(payload[index + 2])
      index += 3
      guard type != 0 else { throw BarnardEventInfoError.zeroTlvType }
      guard type > previousType else { throw BarnardEventInfoError.unorderedTlvTypes }
      previousType = type
      guard length <= payload.count - index else { throw BarnardEventInfoError.truncatedTlv }
      let value = payload.subdata(in: index..<(index + length))
      index += length

      switch type {
      case 0x01:
        guard displayName == nil else { throw BarnardEventInfoError.unorderedTlvTypes }
        displayName = try validatedDisplayName(value)
      case 0x02:
        guard eventCodeHash == nil, value.count == 8 else {
          throw BarnardEventInfoError.invalidEventCodeHash
        }
        eventCodeHash = value
      default:
        continue // Structurally valid extensions are intentionally ignorable.
      }
    }

    guard index == payload.count else { throw BarnardEventInfoError.truncatedTlv }
    guard let displayName else { throw BarnardEventInfoError.missingDisplayName }
    guard let eventCodeHash else { throw BarnardEventInfoError.missingEventCodeHash }
    return BarnardEventInfo(eventDisplayName: displayName, eventCodeHash: eventCodeHash)
  }

  /// A B005 hint is trustworthy only as an unauthenticated hint when it is
  /// internally consistent with the same Peripheral's B004 response.
  public static func matchesB004(_ info: BarnardEventInfo, b004EventCodeHash: Data) -> Bool {
    !b004EventCodeHash.isEmpty && info.eventCodeHash == b004EventCodeHash
  }

  private static func appendLength(_ length: Int, to payload: inout Data) {
    payload.append(UInt8((length >> 8) & 0xff))
    payload.append(UInt8(length & 0xff))
  }

  private static func canonicalDisplayNameBytes(_ displayName: String) throws -> Data {
    let bytes = Data(displayName.utf8)
    // Swift `String ==` uses canonical equivalence, so compare UTF-8 bytes to
    // detect a decomposed spelling rather than accepting it as NFC.
    guard bytes == Data(displayName.precomposedStringWithCanonicalMapping.utf8) else {
      throw BarnardEventInfoError.invalidDisplayName
    }
    guard (1...maximumDisplayNameBytes).contains(bytes.count), containsOnlyPermittedScalars(displayName) else {
      throw BarnardEventInfoError.invalidDisplayName
    }
    return bytes
  }

  private static func validatedDisplayName(_ bytes: Data) throws -> String {
    guard let displayName = String(data: bytes, encoding: .utf8) else {
      throw BarnardEventInfoError.invalidDisplayName
    }
    _ = try canonicalDisplayNameBytes(displayName)
    return displayName
  }

  private static func containsOnlyPermittedScalars(_ value: String) -> Bool {
    value.unicodeScalars.allSatisfy { scalar in
      scalar.value > 0x1f && scalar.value != 0x7f
    }
  }
}

public struct BarnardEventInfoDiscoveryObservation: Equatable {
  public let additionalNamesOmitted: Bool
  public let additionalEventsOmitted: Bool
}

/// Bounded, observer-local retention for one five-minute discovery session.
public final class BarnardEventInfoDiscoverySession {
  private var startedAt: TimeInterval
  private var namesByHash: [Data: Set<Data>] = [:]
  public private(set) var additionalNamesOmitted = false
  public private(set) var additionalEventsOmitted = false
  public var retainedHashCount: Int { namesByHash.count }

  public init(startedAt: TimeInterval) { self.startedAt = startedAt }

  @discardableResult
  public func observe(_ info: BarnardEventInfo, now: TimeInterval) -> BarnardEventInfoDiscoveryObservation {
    if now - startedAt >= 300 {
      startedAt = now
      namesByHash.removeAll()
      additionalNamesOmitted = false
      additionalEventsOmitted = false
    }
    let name = Data(info.eventDisplayName.utf8)
    if var names = namesByHash[info.eventCodeHash] {
      if !names.contains(name) && names.count >= 4 { additionalNamesOmitted = true }
      else { names.insert(name); namesByHash[info.eventCodeHash] = names }
    } else if namesByHash.count >= 32 {
      additionalEventsOmitted = true
    } else {
      namesByHash[info.eventCodeHash] = [name]
    }
    return BarnardEventInfoDiscoveryObservation(
      additionalNamesOmitted: additionalNamesOmitted,
      additionalEventsOmitted: additionalEventsOmitted
    )
  }
}

/// B005's bounded connection/read retry policy. Semantic unavailability is
/// terminal for the session; transport failures get one retry after 30s.
public final class BarnardEventInfoRetryBudget {
  private var attempts: [UUID: Int] = [:]
  private var retryAfter: [UUID: TimeInterval] = [:]
  private var semanticUnavailable: Set<UUID> = []
  private var lastSeen: [UUID: TimeInterval] = [:]

  public init() {}
  public func canStart(_ peer: UUID, now: TimeInterval) -> Bool {
    prune(now: now)
    lastSeen[peer] = now
    return !semanticUnavailable.contains(peer) && (attempts[peer] ?? 0) < 2 && now >= (retryAfter[peer] ?? 0)
  }
  @discardableResult public func recordRecoverableFailure(_ peer: UUID, now: TimeInterval) -> TimeInterval? {
    prune(now: now)
    lastSeen[peer] = now
    let next = (attempts[peer] ?? 0) + 1
    attempts[peer] = next
    guard next < 2 else { return nil }
    let deadline = now + 30
    retryAfter[peer] = deadline
    return deadline
  }
  public func recordSemanticUnavailable(_ peer: UUID, now: TimeInterval = Date().timeIntervalSince1970) {
    prune(now: now)
    lastSeen[peer] = now
    semanticUnavailable.insert(peer)
  }
  public func clear(_ peer: UUID) {
    attempts.removeValue(forKey: peer)
    retryAfter.removeValue(forKey: peer)
    semanticUnavailable.remove(peer)
    lastSeen.removeValue(forKey: peer)
  }
  public func clearAll() {
    attempts.removeAll()
    retryAfter.removeAll()
    semanticUnavailable.removeAll()
    lastSeen.removeAll()
  }
  private func prune(now: TimeInterval) {
    let stalePeers = lastSeen.compactMap { peer, seenAt in
      now >= seenAt && now - seenAt >= 300 ? peer : nil
    }
    stalePeers.forEach(clear)
  }
}

File: packages/android/barnard/src/main/kotlin/org/levarac/barnard/BarnardEventInfo.kt
package org.levarac.barnard

import org.levarac.barnard.BarnardCrypto.toHex

import java.nio.ByteBuffer
import java.nio.CharBuffer
import java.nio.charset.CodingErrorAction
import java.nio.charset.StandardCharsets
import java.text.Normalizer

/** B005's unauthenticated, event-scoped discovery hint. */
public class BarnardEventInfo(
    public val eventDisplayName: String,
    eventCodeHash: ByteArray,
    census: ByteArray? = null,
) {
    private val eventCodeHashBytes: ByteArray = eventCodeHash.copyOf()
    private val censusBytes: ByteArray? = census?.copyOf()

    /** Returns a defensive copy; mutating it cannot affect equality or hashing. */
    public val eventCodeHash: ByteArray
        get() = eventCodeHashBytes.copyOf()

    /** Reserved for the optional 0x10 census extension. v1 leaves it absent. Returns a defensive copy. */
    public val census: ByteArray?
        get() = censusBytes?.copyOf()

    override fun equals(other: Any?): Boolean =
        other is BarnardEventInfo &&
            eventDisplayName == other.eventDisplayName &&
            eventCodeHashBytes.contentEquals(other.eventCodeHashBytes) &&
            censusBytes.contentEquals(other.censusBytes)

    override fun hashCode(): Int =
        31 * (31 * eventDisplayName.hashCode() + eventCodeHashBytes.contentHashCode()) + censusBytes.contentHashCode()
}

/** Kotlin mirror of Swift's [BarnardEventInfoError] reasons. */
public enum class BarnardEventInfoError {
    INVALID_PAYLOAD_LENGTH,
    UNSUPPORTED_FORMAT_VERSION,
    TRUNCATED_TLV,
    ZERO_TLV_TYPE,
    UNORDERED_TLV_TYPES,
    MISSING_DISPLAY_NAME,
    MISSING_EVENT_CODE_HASH,
    INVALID_DISPLAY_NAME,
    INVALID_EVENT_CODE_HASH,
    EVENT_CODE_HASH_MISMATCH,
}

public class BarnardEventInfoException(
    public val reason: BarnardEventInfoError,
) : IllegalArgumentException(reason.name)

/** Local serving state only; neither flag is present in B005 bytes. */
public data class BarnardEventInfoServePolicy(
    val organizerDesignated: Boolean = false,
    val eventActiveForDiscovery: Boolean = false,
) {
    internal val mayServe: Boolean get() = organizerDesignated && eventActiveForDiscovery
}

/** Pure B005 serializer/parser. Hash computation is delegated to B004's function. */
public object BarnardEventInfoCodec {
    public const val FORMAT_VERSION: Int = 1
    public const val MAXIMUM_PAYLOAD_BYTES: Int = 512
    public const val MAXIMUM_DISPLAY_NAME_BYTES: Int = 64

    @JvmStatic
    public fun payloadIfServing(
        policy: BarnardEventInfoServePolicy,
        eventCode: String?,
        eventDisplayName: String?,
        b004EventCodeHash: ByteArray,
    ): ByteArray? {
        if (!policy.mayServe || eventCode.isNullOrEmpty() || eventDisplayName == null) return null
        return serialize(eventCode, eventDisplayName, b004EventCodeHash)
    }

    @JvmStatic
    public fun validateEventDisplayName(eventDisplayName: String) {
        canonicalDisplayNameBytes(eventDisplayName)
    }

    @JvmStatic
    public fun serialize(eventCode: String, eventDisplayName: String, b004EventCodeHash: ByteArray): ByteArray {
        val displayNameBytes = canonicalDisplayNameBytes(eventDisplayName)
        val eventCodeHash = BarnardCrypto.computeEventCodeHash(eventCode)
        requireEventInfo(eventCodeHash.size == 8 && eventCodeHash.contentEquals(b004EventCodeHash), BarnardEventInfoError.EVENT_CODE_HASH_MISMATCH)
        val payload = ByteBuffer.allocate(1 + 3 + displayNameBytes.size + 3 + eventCodeHash.size)
            .put(FORMAT_VERSION.toByte())
            .put(0x01)
            .putShort(displayNameBytes.size.toShort())
            .put(displayNameBytes)
            .put(0x02)
            .putShort(eventCodeHash.size.toShort())
            .put(eventCodeHash)
            .array()
        requireEventInfo(payload.size in 16..MAXIMUM_PAYLOAD_BYTES, BarnardEventInfoError.INVALID_PAYLOAD_LENGTH)
        return payload
    }

    @JvmStatic
    public fun parse(payload: ByteArray): BarnardEventInfo {
        requireEventInfo(payload.size in 16..MAXIMUM_PAYLOAD_BYTES, BarnardEventInfoError.INVALID_PAYLOAD_LENGTH)
        requireEventInfo((payload[0].toInt() and 0xff) == FORMAT_VERSION, BarnardEventInfoError.UNSUPPORTED_FORMAT_VERSION)
        var offset = 1
        var previousType = 0
        var displayName: String? = null
        var eventCodeHash: ByteArray? = null
        while (offset < payload.size) {
            requireEventInfo(offset + 3 <= payload.size, BarnardEventInfoError.TRUNCATED_TLV)
            val type = payload[offset].toInt() and 0xff
            val length = ((payload[offset + 1].toInt() and 0xff) shl 8) or (payload[offset + 2].toInt() and 0xff)
            offset += 3
            requireEventInfo(type != 0, BarnardEventInfoError.ZERO_TLV_TYPE)
            requireEventInfo(type > previousType, BarnardEventInfoError.UNORDERED_TLV_TYPES)
            previousType = type
            requireEventInfo(length <= payload.size - offset, BarnardEventInfoError.TRUNCATED_TLV)
            val value = payload.copyOfRange(offset, offset + length)
            offset += length
            when (type) {
                0x01 -> {
                    requireEventInfo(displayName == null, BarnardEventInfoError.UNORDERED_TLV_TYPES)
                    displayName = validatedDisplayName(value)
                }
                0x02 -> {
                    requireEventInfo(eventCodeHash == null && value.size == 8, BarnardEventInfoError.INVALID_EVENT_CODE_HASH)
                    eventCodeHash = value
                }
            }
        }
        val resolvedDisplayName = displayName ?: throw BarnardEventInfoException(BarnardEventInfoError.MISSING_DISPLAY_NAME)
        val resolvedEventCodeHash = eventCodeHash ?: throw BarnardEventInfoException(BarnardEventInfoError.MISSING_EVENT_CODE_HASH)
        return BarnardEventInfo(resolvedDisplayName, resolvedEventCodeHash)
    }

    @JvmStatic
    public fun matchesB004(info: BarnardEventInfo, b004EventCodeHash: ByteArray): Boolean =
        b004EventCodeHash.isNotEmpty() && info.eventCodeHash.contentEquals(b004EventCodeHash)

    private fun canonicalDisplayNameBytes(displayName: String): ByteArray {
        requireEventInfo(Normalizer.normalize(displayName, Normalizer.Form.NFC) == displayName, BarnardEventInfoError.INVALID_DISPLAY_NAME)
        val bytes = try {
            val encoded = StandardCharsets.UTF_8.newEncoder()
                .onMalformedInput(CodingErrorAction.REPORT)
                .onUnmappableCharacter(CodingErrorAction.REPORT)
                .encode(CharBuffer.wrap(displayName))
            ByteArray(encoded.remaining()).also(encoded::get)
        } catch (_: Exception) {
            throw BarnardEventInfoException(BarnardEventInfoError.INVALID_DISPLAY_NAME)
        }
        requireEventInfo(bytes.size in 1..MAXIMUM_DISPLAY_NAME_BYTES, BarnardEventInfoError.INVALID_DISPLAY_NAME)
        requireEventInfo(displayName.codePoints().allMatch { it > 0x1f && it != 0x7f }, BarnardEventInfoError.INVALID_DISPLAY_NAME)
        return bytes
    }

    private fun validatedDisplayName(bytes: ByteArray): String {
        val displayName = try {
            StandardCharsets.UTF_8.newDecoder()
                .onMalformedInput(CodingErrorAction.REPORT)
                .onUnmappableCharacter(CodingErrorAction.REPORT)
                .decode(ByteBuffer.wrap(bytes))
                .toString()
        } catch (_: Exception) {
            throw BarnardEventInfoException(BarnardEventInfoError.INVALID_DISPLAY_NAME)
        }
        canonicalDisplayNameBytes(displayName)
        return displayName
    }

    private fun requireEventInfo(condition: Boolean, reason: BarnardEventInfoError) {
        if (!condition) throw BarnardEventInfoException(reason)
    }
}

public data class BarnardEventInfoDiscoveryObservation(
    val additionalNamesOmitted: Boolean,
    val additionalEventsOmitted: Boolean,
    val shouldEmitGenericHint: Boolean,
)

/** Replaces an overflowed B005 hint with a marker that has no parsed data. */
internal fun eventInfoForDiscoveryHint(
    eventInfo: BarnardEventInfo,
    shouldEmitGenericHint: Boolean,
): BarnardEventInfo = if (shouldEmitGenericHint) BarnardEventInfo("", ByteArray(0)) else eventInfo

/** Bounded, observer-local state for one five-minute B005 discovery session. */
public class BarnardEventInfoDiscoverySession(startedAtMs: Long) {
    private var startedAtMs = startedAtMs
    private val namesByHash = mutableMapOf<String, MutableSet<String>>()
    public var additionalNamesOmitted: Boolean = false
        private set
    public var additionalEventsOmitted: Boolean = false
        private set
    public val retainedHashCount: Int get() = synchronized(this) { namesByHash.size }

    @Synchronized
    public fun observe(info: BarnardEventInfo, nowMs: Long): BarnardEventInfoDiscoveryObservation {
        if (nowMs - startedAtMs >= 300_000L) {
            startedAtMs = nowMs
            namesByHash.clear()
            additionalNamesOmitted = false
            additionalEventsOmitted = false
        }
        val hash = info.eventCodeHash.toHex()
        val name = info.eventDisplayName
        val names = namesByHash[hash]
        var overflowedThisObservation = false
        when {
            names != null && name !in names && names.size >= 4 -> additionalNamesOmitted = true
            names != null -> names += name
            namesByHash.size >= 32 -> {
                additionalEventsOmitted = true
                overflowedThisObservation = true
            }
            else -> namesByHash[hash] = mutableSetOf(name)
        }
        return BarnardEventInfoDiscoveryObservation(
            additionalNamesOmitted,
            additionalEventsOmitted,
            overflowedThisObservation,
        )
    }
}

/** Two B005 transport attempts per peer/session, with a fixed 30-second retry delay. */
public class BarnardEventInfoRetryBudget {
    private val attempts = mutableMapOf<String, Int>()
    private val retryAfterMs = mutableMapOf<String, Long>()
    private val semanticUnavailable = mutableSetOf<String>()
    private val lastSeenMs = mutableMapOf<String, Long>()

    @Synchronized
    public fun canStart(peer: String, nowMs: Long): Boolean {
        prune(nowMs)
        lastSeenMs[peer] = nowMs
        return peer !in semanticUnavailable && (attempts[peer] ?: 0) < 2 && nowMs >= (retryAfterMs[peer] ?: 0L)
    }

    @Synchronized
    public fun recordRecoverableFailure(peer: String, nowMs: Long): Long? {
        prune(nowMs)
        lastSeenMs[peer] = nowMs
        val count = (attempts[peer] ?: 0) + 1
        attempts[peer] = count
        if (count >= 2) return null
        return (nowMs + 30_000L).also { retryAfterMs[peer] = it }
    }

    /** A successful B005 read still consumes one of the two session attempts. */
    @Synchronized
    public fun recordSuccessfulAttempt(peer: String, nowMs: Long) {
        prune(nowMs)
        lastSeenMs[peer] = nowMs
        attempts[peer] = (attempts[peer] ?: 0) + 1
        retryAfterMs.remove(peer)
    }

    @Synchronized
    public fun recordSemanticUnavailable(peer: String, nowMs: Long = System.currentTimeMillis()) {
        prune(nowMs)
        lastSeenMs[peer] = nowMs
        semanticUnavailable += peer
    }

    @Synchronized
    public fun clear(peer: String) {
        attempts.remove(peer)
        retryAfterMs.remove(peer)
        semanticUnavailable.remove(peer)
        lastSeenMs.remove(peer)
    }

    @Synchronized
    public fun clearAll() {
        attempts.clear()
        retryAfterMs.clear()
        semanticUnavailable.clear()
        lastSeenMs.clear()
    }

    private fun prune(nowMs: Long) {
        val stalePeers = lastSeenMs.filterValues { seenAt -> nowMs >= seenAt && nowMs - seenAt >= 300_000L }.keys
        stalePeers.forEach(::clear)
    }
}
